batch.go

```diff
@@ -0,0 +1,197 @@
+		b.add(entry)
+	}
+
+	return b
+}
+
+// add an entry to the batch
+func (b *Batch) add(entry Entry) {
+	b.Bytes += len(entry.Line)
+
+	// Append the entry to an already existing stream (if any)
+	labels := entry.Labels.String()
+	if stream, ok := b.Streams[labels]; ok {
+		stream.Entries = append(stream.Entries, entry.Entry)
+		return
+	}
+
+	// Add the entry as a new stream
+	b.Streams[labels] = &logproto.Stream{
+		Labels:  labels,
+		Entries: []logproto.Entry{entry.Entry},
+	}
+}
+
+// sizeBytes returns the current batch size in bytes
+func (b *Batch) sizeBytes() int {
+	return b.Bytes
+}
+
+// sizeBytesAfter returns the size of the batch after the input entry
+// will be added to the batch itself
+func (b *Batch) sizeBytesAfter(entry Entry) int {
+	return b.Bytes + len(entry.Line)
+}
+
+// age of the batch since its creation
+func (b *Batch) age() time.Duration {
+	return time.Since(b.CreatedAt)
+}
+
+// encodeSnappy encodes the batch as snappy-compressed push request, and
+// returns the encoded bytes and the number of encoded entries
+func (b *Batch) encodeSnappy() ([]byte, int, error) {
+	req, entriesCount := b.createPushRequest()
+	buf, err := proto.Marshal(req)
+	if err != nil {
+		return nil, 0, err
+	}
+	buf = snappy.Encode(nil, buf)
+	return buf, entriesCount, nil
+}
+
+// encodeJSON encodes the batch as JSON push request, and returns the encoded
+// bytes and the number of encoded entries
+func (b *Batch) encodeJSON() ([]byte, int, error) {
+	req, entriesCount := b.createPushRequest()
+	buf, err := json.Marshal(req)
+	if err != nil {
+		return nil, 0, err
+	}
+	return buf, entriesCount, nil
+}
+
+// createPushRequest creates a push request and returns it, together with
+// number of entries
+func (b *Batch) createPushRequest() (*logproto.PushRequest, int) {
+	req := logproto.PushRequest{
+		Streams: make([]logproto.Stream, 0, len(b.Streams)),
+	}
+
+	entriesCount := 0
+	for _, stream := range b.Streams {
+		req.Streams = append(req.Streams, *stream)
+		entriesCount += len(stream.Entries)
+	}
+	return &req, entriesCount
+}
+
+// generateEntries creates a list of log entries
+func generateEntries(ctx context.Context, tenantID string, pool LabelPool, s int, min int, max int) []Entry {
+	state := lib.GetState(ctx)
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "localhost"
+	}
+
+	result := make([]Entry, 0)
+	var labels model.LabelSet
+	var n int
+	// create `[min,max)` entries for `s` streams
+	for i := 0; i < s; i++ {
+		labels = labelsFromPool(pool)
+		labels[model.InstanceLabel] = model.LabelValue(fmt.Sprintf("vu%d.%s.local", state.VUID, hostname))
+
+		n = min + rand.Intn(max-min)
+		entries := make([]Entry, n)
+		for i := 0; i < n; i++ {
+			entries[i] = Entry{
+				Entry: logproto.Entry{
+					Timestamp: time.Now(),
+					Line:      flog.NewLog(string(labels[model.LabelName("format")]), time.Now()),
+				},
+				TenantID: tenantID,
+				Labels:   labels,
+			}
+			time.Sleep(10 * time.Millisecond)
```

Comment:
I am not sure that we have to sleep during the data generation. I would add a delay before sending the batch. why do we need it here?